subsys/logging/log_output_syst.c

```diff
@@ -786,27 +786,97 @@ void log_output_msg2_syst_process(const struct log_output *output,
 	uint8_t *data = log_msg2_get_package(msg, &len);
 
 	if (len) {
-		char *buf = data, *fmt;
+		/*
+		 * The package format:
+		 *
+		 * [0]: Argument list size including package header
+		 *      (sizeof(uint8_t *)) and format string pointer,
+		 *      in number of uint32_t words.
+		 * [1]: Number of appened strings in package.
+		 *
+		 * [sizeof(uint8_t *)]: Pointer to format string.
+		 *                      Can be NULL so needs further
+		 *                      processing.
+		 *
+		 * [sizeof(uint8_t *) * 2]: Start of arguments.
+		 *
+		 * [[0] * sizeof(uint32_t)]: Start of appended strings.
+		 */
+
+		uint8_t num_strings = data[1];
+		char *output_buf = data;
+		char *fmt = *(char **)(data + sizeof(uint8_t *));
+
+		if (num_strings != 0U) {
+			if (len > CONFIG_LOG_MIPI_SYST_CBSTR_BUF_LEN) {
+				output_buf = NULL;
+			} else {
+				/*
+				 * There are appended strings in the package. So
+				 * the string pointers in the argument list
+				 * need to be replaced with the pointers
+				 * in the append strings list.
+				 */
+
+				uint32_t remaining = num_strings;
+				uint8_t cb_buf[CONFIG_LOG_MIPI_SYST_CBSTR_BUF_LEN];
+				uint32_t *cb_buf_u32 = (uint32_t *)cb_buf;
+				uint32_t cb_sz = (uint32_t)data[0] * sizeof(uint32_t);
+
+				/* Make a copy since we need to manipulate the arg list */
+				(void)memcpy(cb_buf, data, len);
+
+				/* Appended strings are after the argument list */
+				uint8_t *ap_strs = cb_buf + cb_sz;
+
+				while (remaining > 0U) {
+					/*
+					 * The first byte of each appended string points
+					 * to the location of the argument in the form
+					 * of n-th uint32_t word in the packaged string
+					 * buffer, followed by the null-terminated string.
+					 */
+					uintptr_t arg_ptr = (uintptr_t)&cb_buf_u32[ap_strs[0]];
+					uintptr_t str_ptr = (uintptr_t)(++ap_strs);
+
+					/* Replace the argument with the string pointer */
+					*(char **)arg_ptr = (char *)str_ptr;
+
+					/* Skip to the end of this string */
+					while (*(ap_strs++) != 0U) {
+					};
+
+					remaining--;
+				}
+
+				/*
+				 * The output_buf contains the updated argument list.
+				 * So use this for output.
+				 */
+				output_buf = cb_buf;
+			}
+		}
 
-		fmt = ((char **)buf)[1];
-		buf += sizeof(char *) * 2;
+		if (output_buf != NULL) {
```

Comment:
There are corner cases for preparing va_list (xtensa, x86_64), see set of `cbprintf_via_va_list` functions: https://github.com/zephyrproject-rtos/zephyr/blob/fdc25cd44c2b693aecff96d3dfcd0ee088d48d5b/lib/os/cbprintf_packaged.c#L75

Given that maybe instead of `cbprintf_locate/settle` approach, it would be better to have something like `cbpprintf_external(ext_formatter func, void *ctx, void *packaged)` where external formatter would be function like `foo(void *ctx, const char *fmt, va_list ap)`.

Reply:
I have updated `cbprintf_packaged.c` to use external formatter. Hopefully this aligns with what you are suggesting.